sdk/dotnet/Pulumi/Testing/MockMonitor.cs

```diff
@@ -0,0 +1,123 @@
+using Pulumirpc;
+
+namespace Pulumi.Testing
+{
+    internal class MockMonitor : IMonitor
+    {
+        private readonly IMocks _mocks;
+        private readonly Serializer _serializer = new Serializer();
+        private string? _rootResourceUrn = null;
+        
+        public readonly List<string> Errors = new List<string>();
+        public readonly List<Resource> Resources = new List<Resource>();
+
+        public ImmutableDictionary<string, object> StackOutputs { get; private set; } =
+            ImmutableDictionary<string, object>.Empty;
+
+        public MockMonitor(IMocks mocks)
+        {
+            _mocks = mocks;
+        }
+
+        public async Task<InvokeResponse> InvokeAsync(InvokeRequest request)
+        {
+            var result = await _mocks.CallAsync(request.Tok, ToDictionary(request.Args), request.Provider);
+            return new InvokeResponse { Return = await SerializeAsync(result) };
+        }
+
+        public async Task<ReadResourceResponse> ReadResourceAsync(Resource resource, ReadResourceRequest request)
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name,
+                ToDictionary(request.Properties), request.Provider, request.Id);
+            this.Resources.Add(resource);
+            return new ReadResourceResponse
+            {
+                Urn = NewUrn(request.Parent, request.Type, request.Name),
+                Properties = await SerializeAsync(state)
+            };
+        }
+
+        public async Task<RegisterResourceResponse> RegisterResourceAsync(Resource resource, RegisterResourceRequest request)
+        {
+            var (id, state) = await _mocks.NewResourceAsync(request.Type, request.Name, ToDictionary(request.Object),
+                request.Provider, request.ImportId);
+            this.Resources.Add(resource);
+            return new RegisterResourceResponse
+            {
+                Id = id ?? request.ImportId,
+                Urn = NewUrn(request.Parent, request.Type, request.Name),
+                Object = await SerializeAsync(state)
+            };
+        }
+
+        public Task RegisterResourceOutputsAsync(RegisterResourceOutputsRequest request)
+        {
+            var stackUrn = NewUrn("", Stack._rootPulumiStackTypeName, $"{Deployment.Instance.ProjectName}-{Deployment.Instance.StackName}");
+            if (request.Urn == stackUrn)
+            {
+                StackOutputs = ToDictionary(request.Outputs);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        public Task LogAsync(LogRequest request)
+        {
+            if (request.Severity == LogSeverity.Error)
+            {
+                this.Errors.Add(request.Message);
+            }
+            
+            return Task.CompletedTask;
+        }
+
+        public Task<SetRootResourceResponse> SetRootResourceAsync(SetRootResourceRequest request)
+        {
+            _rootResourceUrn = request.Urn;
```

Comment:
make threadsafe, bail if we get mltiple sets of this.

Reply:
added